module-system/sov-modules-api/src/gas.rs

```diff
@@ -0,0 +1,158 @@
+pub mod sov_modules_api {
+    use anyhow::Result;
+
+    use core::marker::PhantomData;
+
+    // We add `remainder_funds` to `WorkingSet` as a simple `u64`; the gas will already be converted
+    // to scalar value when charged from `WorkingSet`.
+
+    pub trait GasUnit {
+        //: serde::Serialize + serde::de::DeserializeOwned {
+        fn value(&self, price: &Self) -> u64;
+    }
+
+    pub trait Context {
+        type GasUnit: GasUnit;
+    }
+
+    pub struct GasoMeter<GasUnit> {
+        remaining_funds: u64,
+        gas_price: GasUnit,
+    }
+
+    impl<GU: GasUnit> GasoMeter<GU> {
+        pub fn new(remaining_funds: u64, gas_price: GU) -> Self {
+            Self {
+                remaining_funds,
+                gas_price,
+            }
+        }
+
+        pub fn charge_gas(&mut self, gas: &GU) -> Result<()> {
+            // TODO add check
+            self.remaining_funds = self
+                .remaining_funds
+                .saturating_sub(gas.value(&self.gas_price));
+
+            Ok(())
+        }
+    }
+    pub struct WorkingSet<C: Context> {
+        gaso_meter: GasoMeter<C::GasUnit>,
+        _context: PhantomData<C>,
+    }
+
+    impl<C: Context> WorkingSet<C> {
+        pub fn new(gaso_meter: GasoMeter<C::GasUnit>) -> Self {
+            Self {
+                gaso_meter,
+                _context: PhantomData,
+            }
+        }
+
+        pub fn charge_gas(&mut self, gas: &C::GasUnit) -> Result<()> {
+            self.gaso_meter.charge_gas(gas)
+        }
+    }
+
+    #[derive(serde::Serialize, serde::Deserialize)]
+    pub struct TupleGasUnit {
+        pub a: u64,
+        pub b: u64,
+        pub c: u64,
+    }
+
+    impl GasUnit for TupleGasUnit {
+        fn value(&self, price: &Self) -> u64 {
+            self.a.saturating_mul(price.a).saturating_add(
+                self.b
+                    .saturating_mul(price.b)
+                    .saturating_add(self.c.saturating_mul(price.c)),
+            )
+        }
+    }
+
+    pub struct DefaultContext {}
+
+    impl Context for DefaultContext {
+        type GasUnit = TupleGasUnit;
+    }
+
+    // The `GasConfig` will have the whole implementation encapsulated by macros.
+
+    pub trait GasConfig {
+        const MANIFEST: Self;
+    }
+
+    pub trait Module {
+        type Context: Context;
+        type GasConfig: GasConfig;
+    }
+}
+
+pub mod foo {
+    use std::marker::PhantomData;
+
+    use anyhow::Result;
+
+    use super::sov_modules_api::{Context, GasConfig, GasUnit, Module, WorkingSet};
+
+    //#[derive(serde::Serialize, serde::Deserialize)]
+    pub struct FooGasConfig<GS> {
+        complex_math_operation: GS,
+        another_operation: GS,
+    }
+
+    impl<GS: GasUnit> GasConfig for FooGasConfig<GS> {
+        const MANIFEST: Self = {
+            Self {
+                complex_math_operation: todo!(),
```

Comment:
I'm unsure about this part; how we should read the manifest?

Reply:
It will be a macro expansion executed at compilation